django/contrib/auth/views.py

```diff
@@ -446,12 +454,25 @@ def dispatch(self, *args, **kwargs):
         self.validlink = False
         self.user = self.get_user(kwargs['uidb64'])
 
-        if self.user is not None and self.token_generator.check_token(self.user, kwargs['token']):
-            self.validlink = True
-        else:
-            return self.render_to_response(self.get_context_data())
-
-        return super(PasswordResetConfirmView, self).dispatch(*args, **kwargs)
+        if self.user is not None:
+            token = kwargs['token']
+            if token == INTERNAL_RESET_URL_TOKEN:
+                # If the token is valid, display the form & allow a password reset.
+                session_token = self.request.session.get(INTERNAL_RESET_SESSION_TOKEN)
+                if self.token_generator.check_token(self.user, session_token):
+                    self.validlink = True
+                    return super(PasswordResetConfirmView, self).dispatch(*args, **kwargs)
+            else:
+                if self.token_generator.check_token(self.user, token):
+                    # Store the token in the session, to not leak it in a referer
+                    # and redirect to a token-free URL.
+                    # See https://robots.thoughtbot.com/is-your-site-leaking-password-reset-links for details.
+                    self.request.session[INTERNAL_RESET_SESSION_TOKEN] = token
+                    return HttpResponseRedirect(self._build_redirect_url())
+
+        # Otherwise delete it immediately and render an error page.
+        self.request.session.pop(INTERNAL_RESET_SESSION_TOKEN, None)
```

Comment:
I think that can get deleted now, we only set the token in the session if it is valid and then we do not reach that part?